src/dswx_sar/pre_processing.py

```diff
@@ -0,0 +1,407 @@
+    '''
+    array1 = np.asarray(array1, dtype='float32')
+    array2 = np.asarray(array2, dtype='float32')
+
+    return array1 / array2
+
+def pol_coherence(array1, array2, array3):
+    '''
+    Compute polarimetric coherence from two co-pol and one cross-pol
+    '''
+    array1 = np.asarray(array1, dtype='float32') # VVVV
+    array2 = np.asarray(array2, dtype='float32') # VHVH
+    array3 = np.asarray(array3, dtype='complex') # VVVH
+
+    return np.abs(array3 / np.sqrt(array1 * array2))
+
+class AncillaryRelocation:
+    '''
+    Relocate the ancillary file to have same geographic extents and 
+    spacing as the given rtc file.
+    '''
+    def __init__(self, rtc_file_name, scratch_dir):
+        """Initialized AncillaryRelocation Class with rtc_file_name
+
+        Parameters
+        ----------
+        rtc_file_name : str
+            file name of RTC input HDF5
+        scratch_dir : str
+            scratch directory to save the relocated file
+        """
+
+        self.rtc_file_name = rtc_file_name
+        if  h5py.is_hdf5(rtc_file_name):
+            self.xcoord_rtc, self.ycoord_rtc = dswx_sar_util.read_rtc_lat_lon(rtc_file_name)
+            with h5py.File(rtc_file_name, 'r') as f4:
+                self.epsg = int(np.array(f4['/science/LSAR/GCOV/grids/frequencyA/projection']))
+        else:
+            self.ycoord_rtc, self.xcoord_rtc = self.read_x_y_array_geotiff(rtc_file_name)
+            reftif = gdal.Open(rtc_file_name)
+            proj = osr.SpatialReference(wkt=reftif.GetProjection())
+            self.epsg = proj.GetAttrValue('AUTHORITY',1)
+            reftif = None
+        self.scratch_dir = scratch_dir
+        print('ancillary_reloation', self.epsg)
+
+    def _relocate(self,
+                  ancillary_file_name,
+                  relocated_file_str,
+                  method='near'):
+
+        """ resample image
+
+        Parameters
+        ----------
+        ancillary_file_name : str
+            file name of ancilary data
+        relocated_file_str : str
+            file name of output
+        method : str
+            interpolation method
+
+        Returns
+        -------
+        resampled_slc or slc_demodulate: numpy.ndarray
+            numpy array of bandpassed slc
+            if resampling is True, return resampled slc with bandpass and demodulation
+            if resampling is False, return slc with bandpass and demodulation without resampling
+        meta : dict
+            dict containing meta data of bandpassed slc
+            center_frequency, rg_bandwidth, range_spacing, slant_range
+        """
+
+        self.interpolate_gdal(str(self.rtc_file_name),
+                         ancillary_file_name,
+                         os.path.join(self.scratch_dir, relocated_file_str),
+                         method)
+
+    def interpolate_gdal(self, ref_file, input_tif_str, output_tif_str, method, epsg=None):
+        print(f"> gdalwarp {input_tif_str} -> {output_tif_str}")
+
```

Comment:
This function seems to to something complex. Please add a brief docstring to explain what it does.

Reply:
I added the docstring. Thank you